elpy/rpc.py

```diff
@@ -67,14 +67,18 @@ def read_json(self):
             raise EOFError()
         return json.loads(line)
 
+    def json_defaults(self, x):
+        if isinstance(x, Path):
+            return str(x)
+
     def write_json(self, **kwargs):
         """Write an JSON object on a single line.
 
         The keyword arguments are interpreted as a single JSON object.
         It's not possible with this method to write non-objects.
 
         """
-        self.stdout.write(json.dumps(kwargs) + "\n")
+        self.stdout.write(json.dumps(kwargs, default=self.json_defaults) + "\n")
```

Comment:
The docs for [`json.dumps` ](https://docs.python.org/3.9/library/json.html#json.dump) say that it should raise a `TypeError` if the object cannot be serializable, (In this case, all other objects that are not `Path` and are not primitive types will be encoded as `null` since `json_defaults` is implicitly returning `None`.

```python
>>> import json
>>> json.dumps(None)
'null'
>>> 
```
Also relevant [CPython source code](https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/json/encoder.py#L413)  that shows this will happen

I don't think this is the intended outcome?

Reply:
> The docs for [`json.dumps` ](https://docs.python.org/3.9/library/json.html#json.dump) say that it should raise a `TypeError` if the object cannot be serializable, (In this case, all other objects that are not `Path` and are not primitive types will be encoded as `null` since `json_defaults` is implicitly returning `None`.
> 
> ```python
> >>> import json
> >>> json.dumps(None)
> 'null'
> >>> 
> ```
> 
> Also relevant [CPython source code](https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/json/encoder.py#L413) that shows this will happen
> 
> I don't think this is the intended outcome?

Hi @gopar. Sorry I've been extremely busy. Yes, definitely that shouldn't be there. I had completely overlooked that and forgot to return the default value. It should be:

```python
def json_defaults(x):
    if isinstance(x, Path): 
         return str(x) 
    else: 
        return json.dumps(x)
```
See docstring of [default](https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/json/encoder.py#L160) in `JSONEncoder`.

Actually `pydantic` would be a good approach to go for it as it is primarily a parsing library. Though I think @gfederix's PR (See discussion in https://github.com/jorgenschaefer/elpy/pull/1895) requires a few fixes.

At this point I'm not sure which things to refactor, for example in `rpc.py`, as there are parallel changes in  https://github.com/jorgenschaefer/elpy/pull/1895 which are not yet merged. Any suggestions?